ui/src/views/Documentation/index.jsx

```diff
@@ -106,63 +182,220 @@ export default class Documentation extends Component {
+      ) {
+        return process.env.TASKCLUSTER_ROOT_URL + schemaId;
+      }
+
+      return `https://schemas.taskcluster.net/${schemaId.replace(
+        /^\/schemas\//,
+        ''
+      )}`;
+    }
+
+    return schemaId;
+  }
+
+  sanitizeSchema(schema) {
+    if (schema.$id) {
+      return {
+        ...schema,
+        $id: this.buildSchemaId(schema.$id),
+      };
+    }
+
+    if (schema.id) {
+      return {
+        ...schema,
+        id: this.buildSchemaId(schema.id),
+      };
+    }
+
+    return schema;
+  }
+
+  // Returns a mapping between the HTML element and the desired component
+  components() {
+    return {
+      a: this.anchorFactory,
+      h1: this.headingFactory('h1'),
+      h2: this.headingFactory('h2'),
+      h3: this.headingFactory('h3'),
+      h4: this.headingFactory('h4'),
+      h5: this.headingFactory('h5'),
+      h6: this.headingFactory('h6'),
+      img: this.imageFactory,
+    };
+  }
+
   async load() {
     try {
-      const { default: Document } = await this.handleImport(
-        this.props.match.params.path
+      const { params } = this.props.match;
+      const pathname = params.path || 'index';
+      const isJsonFile =
+        pathname.startsWith('reference') &&
+        (pathname.endsWith('api') || pathname.endsWith('events'));
+      const { loader, path } = readDocFile(
+        `${pathname}.${isJsonFile ? 'json' : 'md'}`
       );
+      const { default: Page } = await loader;
+
+      absolutePath = path;
+
+      // use `isMDXComponent` once https://github.com/mdx-js/mdx/pull/369 is merged
+      if (!Page.prototype) {
+        const entries = await this.getReferenceEntries(Page.entries);
+
+        return this.setState({
+          Page: null,
+          pageInfo: null,
+          error: null,
+          referenceJson: Object.assign({}, Page, { entries }),
+        });
+      }
+
       const pageInfo = this.getPageInfo();
 
-      this.setState({ Document, pageInfo, error: null });
+      this.setState({ Page, pageInfo, error: null, referenceJson: null });
     } catch (error) {
       this.setState({ error });
     }
   }
 
-  async handleImport(url) {
-    const doc = url ? url.replace(/\/$/, '') : 'index';
+  getDocument(Page) {
+    if (!Page) {
+      return null;
+    }
 
-    return import(/* webpackChunkName: 'Documentation.page' */ `../../docs/${doc}.md`).catch(
-      () =>
-        import(/* webpackChunkName: 'Documentation.page' */ `../../docs/${doc}/index.md`)
+    const page = renderToString(
+      <BrowserRouter>
+        <Page components={this.components()} />
+      </BrowserRouter>
     );
+
+    return page
+      .split(DOCS_SCHEMA_REGEX)
+      .map(html =>
+        !isUrl(html) && html.startsWith('<') ? (
+          html
+        ) : (
+          <SchemaTable key={html} schema={html} />
+        )
+      );
   }
 
+  getReferenceEntries = entries => {
+    const projectName = this.props.match.params.path.split('/')[2];
```

Comment:
I'd prefer to not try to parse paths like this -- best to get this from some metadata attached to the object when it was created.  Otherwise, if we move things around then suddenly assumptions like this start generating unexpected behavior.